x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/sp/SamlServiceProviderDocument.java

```diff
@@ -0,0 +1,404 @@
+
+    public static class Privileges {
+        @Nullable
+        public String application;
+        public String resource;
+        @Nullable
+        public String loginAction;
+        public Map<String, String> groupActions;
+
+        public void setApplication(String application) {
+            this.application = application;
+        }
+
+        public void setResource(String resource) {
+            this.resource = resource;
+        }
+
+        public void setLoginAction(String loginAction) {
+            this.loginAction = loginAction;
+        }
+
+        public void setGroupActions(Map<String, String> groupActions) {
+            this.groupActions = groupActions;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            final Privileges that = (Privileges) o;
+            return Objects.equals(application, that.application) &&
+                Objects.equals(resource, that.resource) &&
+                Objects.equals(loginAction, that.loginAction) &&
+                Objects.equals(groupActions, that.groupActions);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(application, resource, loginAction, groupActions);
+        }
+    }
+
+    public static class AttributeNames {
+        public String principal;
+        @Nullable
+        public String email;
+        @Nullable
+        public String name;
+        @Nullable
+        public String groups;
+
+        public void setPrincipal(String principal) {
+            this.principal = principal;
+        }
+
+        public void setEmail(String email) {
+            this.email = email;
+        }
+
+        public void setName(String name) {
+            this.name = name;
+        }
+
+        public void setGroups(String groups) {
+            this.groups = groups;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            final AttributeNames that = (AttributeNames) o;
+            return Objects.equals(principal, that.principal) &&
+                Objects.equals(email, that.email) &&
+                Objects.equals(name, that.name) &&
+                Objects.equals(groups, that.groups);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(principal, email, name, groups);
+        }
+    }
+
+    public String docId;
+
+    public String name;
+
+    public String entityId;
+
+    public String acs;
+
+    public boolean enabled = true;
+    public Instant created;
+    public Instant lastModified;
+
+    @Nullable
+    public String nameIdFormat;
+
+    @Nullable
+    public Long authenticationExpiryMillis;
+
+    public List<String> signingCertificates = List.of();
+
```

Comment:
How would that work? We you proposing that we poll for metadata automatically? I wasn't expecting that we'd do that (at least, not now)

Reply:
I meant that we can support signing the IDP metadata with a different key, depending on who the SP that will consume/validate them ( For the same reasons that we support different keys for SAML message signing , i.e. so that weird/specific requirements from SPs will not force us to rotate signing keys for all SPs )